Push/PushCrypto.swift

```diff
@@ -113,22 +115,123 @@ extension PushCrypto {
     }
 
     func aesgcm(ciphertext: Data, decryptWith rawRecvPrivKey: Data, authenticateWith authSecret: Data, encryptionHeader: String, cryptoKeyHeader: String) throws -> Data {
-        var plaintextLen = ece_aesgcm_plaintext_max_length(ciphertext.count) + 1
+        let saltLength = Int(ECE_SALT_LENGTH)
+        var salt = [UInt8](repeating: 0, count: saltLength)
+
+        let rawSenderPubKeyLength = Int(ECE_WEBPUSH_PUBLIC_KEY_LENGTH)
+        var rawSenderPubKey = [UInt8](repeating: 0, count: rawSenderPubKeyLength)
+
+        var rs = UInt32(0)
+
+        let paramsErr = ece_webpush_aesgcm_headers_extract_params(
+                    cryptoKeyHeader, encryptionHeader,
+                    &salt, saltLength,
+                    &rawSenderPubKey, rawSenderPubKeyLength,
+                    &rs)
+        if paramsErr != ECE_OK {
+            throw PushCryptoError.decryptionError(errCode: paramsErr)
+        }
+
+        var plaintextLen = ece_aesgcm_plaintext_max_length(rs, ciphertext.count) + 1
         var plaintext = [UInt8](repeating: 0, count: plaintextLen)
 
-        let err = ece_webpush_aesgcm_decrypt(
+        let decryptErr = ece_webpush_aesgcm_decrypt(
                 rawRecvPrivKey.getBytes(), rawRecvPrivKey.count,
                 authSecret.getBytes(), authSecret.count,
-                cryptoKeyHeader, encryptionHeader,
+                &salt, salt.count,
+                &rawSenderPubKey, rawSenderPubKey.count,
+                rs,
                 ciphertext.getBytes(), ciphertext.count,
                 &plaintext, &plaintextLen)
 
-        if err != ECE_OK {
-            throw PushCryptoError.decryptionError(errCode: err)
+        if decryptErr != ECE_OK {
+            throw PushCryptoError.decryptionError(errCode: decryptErr)
         }
 
         return Data(bytes: plaintext, count: plaintextLen)
     }
+
+    func aesgcm(plaintext: String, encryptWith rawRecvPubKey: String, authenticateWith authSecret: String, rs: Int, padLen: Int) throws -> (encryptionHeader: String, cryptoKeyHeader: String, ciphertext: String) {
```

Comment:
Good idea, thanks! I see the two overloads return `ciphertext` as `Data` or a `String`, so I'm not sure if it's worth adding a type for the tuple, but I factored out `encryptionHeader` and `cryptoKeyHeader` into their own struct.